### PR TITLE
add stack checks

### PR DESCRIPTION
**Pull request by saaramar**

### CODE REVIEW

**Comment by rmn30** [comment]
I note that this doesn't test with a sealed stack pointer. This is harder to arrange because you would need a sealing cap to seal the stack pointer with or you could use a sentry but then it would have incorrect permissions too. I think it would fail in the same way as an untagged stack pointer anyway so providing that works it is probably fine.

File: sdk/core/switcher/entry.S [hunk 1]
```diff
@@ -91,11 +91,21 @@ compartment_switcher_entry:
...
+	clhu			tp, TrustedStack_offset_frameoffset(ct2)
+	cgetlen         t2, ct2
+	bgeu           	tp, t2, .Lout_of_trusted_stack
+	// make sure the caller's CSP is valid
+	cgettag         t2, csp
+	beqz            t2, .Lforce_unwind
+	// make sure the caller's CSP has the expected permissions
+	cgetperm        t2, csp
+	li              tp, 0x7e
+	bne	            tp, t2, .Lforce_unwind
```
Comment (davidchisnall) [inline sdk/core/switcher/entry.S:105]: Can these be moved before we access mscratch?  We don't need access to any privileged state to do these checks.

File: sdk/core/switcher/entry.S [hunk 2]
```diff
@@ -415,6 +425,13 @@ exception_entry_asm:
 	li          a1, 0
 	// If we don't have enough space, give up and force unwind
 	bltu        t1, tp, .Lforce_unwind
+	// make sure the caller's CSP is valid
+	cgettag     t1, ct0
+	beqz        t1, .Lforce_unwind
+	// make sure the caller's CSP has the expected permissions
+	cgetperm    t1, ct0
+	li 	        tp, 0x7e
+	bne	        tp, t1, .Lforce_unwind
```
Comment (davidchisnall) [inline sdk/core/switcher/entry.S:434]: These checks are the same as above, can you make them a macro? We should also add a sealing check.

Comment (saaramar) [reply sdk/core/switcher/entry.S:434]: Great point, thanks!

File: sdk/core/switcher/entry.S (hunk 1 above)
Comment (saaramar) [reply sdk/core/switcher/entry.S:105]: True, great point. I moved it to the beginning of the trap

File: sdk/core/switcher/entry.S [hunk 3]
```diff
@@ -79,23 +79,47 @@ switcher_scheduler_entry_csp:
...
 #ifndef NDEBUG
 	// XXX: This line is useless, only for mscratch to show up in debugging.
 	cmove			ct2, ct2
 #endif
-	clhu			tp, TrustedStack_offset_frameoffset(ct2)
+
 	// make sure the trusted stack is still in bounds
-	cgetlen                 t2, ct2
-	bgeu                    tp, t2, .Lout_of_trusted_stack
+	clhu			tp, TrustedStack_offset_frameoffset(ct2)
```
Comment (davidchisnall) [inline sdk/core/switcher/entry.S:117]: Can you fix indenting? It looks like you're mixing tabs and spaces.

File: sdk/core/switcher/trusted-stack-assembly.h [hunk 4]
```diff
@@ -44,3 +44,5 @@ EXPORT_ASSEMBLY_SIZE(TrustedStackFrame, (7 * 8))
 
 #define TSTACKOFFSET_FIRSTFRAME                                                \
 	(TrustedStack_offset_frameoffset + TSTACK_HEADER_SZ)
+
+#define COMPARTMENT_STACK_PERMISSIONS 0x7e
```
[the remark this answers is not in the document]
Comment (saaramar) [reply sdk/core/switcher/trusted-stack-assembly.h:48]: Oh right, thanks! Done

File: sdk/core/switcher/entry.S (hunk 3 above)
Comment (saaramar) [reply sdk/core/switcher/entry.S:117]: good catch, thanks :)

**Review by davidchisnall — APPROVED** [review]

**Pull request merged**